src/partial-result-stream.ts

```diff
@@ -355,13 +359,25 @@ export function partialResultStream(
     },
   });
 
+  // This listener ensures that the last request that executed successfully
+  // after one or more retries will end the requestsStream.
+  const endListener = () => {
+    if (lastErr) {
+      setImmediate(() => requestsStream.end());
```

Comment:
I don't quite understand this. Why do we need to do this? `endListener` is triggered by an `end` event in `lastRequestStream`, which means this request is successfully executed. Then, we can close the entire `requestsStream` (a merged stream). Why don't we need to close `requestsStream` in the previous code but now we do?

Reply:
Yes, this was the part that took me hours to figure out as well. The basic answer is: The original solution seems to never have worked as intended in the first place.

The problem is that the first request stream that causes an error will never indicate to the merged `requestsStream` that it has ended. This can be illustrated with the following test case (not added to the code, as this test case will actually fail):

```typescript
    it('should end merged stream', done => {
      // This test case will never finish when this flag is set to true.
      const simulateRetryAfterError = true;
      const mergedStream = mergeStream();

      eventsIntercept.patch(mergedStream);
      // tslint:disable-next-line no-any
      (mergedStream as any).intercept('error', err => {
        // By intercepting the error, the underlying input stream that caused
        // the error will never end. That in turn will cause the merged stream
        // also to never end.
        console.log('intercepted error');
      });

      const inputStream1 = new Readable();
      inputStream1._read = (size: number): void => {
        setImmediate(() => {
          inputStream1.emit('data', 'one');
          inputStream1.emit('data', 'two');
          // inputStream1.emit('end'); // Uncomment this line and comment out the next line to make the test case work.
          inputStream1.emit('error', new Error('test error'));
        });
      };
      inputStream1.on('end', () => console.log('Input stream 1 ended'));
      if (simulateRetryAfterError) {
        // Add a stream that ends with an error. That error is then intercepted by the merged
        // stream, which causes the merged stream to think that the underlying
        // stream has never ended, which in turn prevents the mergedStream from
        // finishing.
        mergedStream.add(inputStream1);
      }

      const inputStream2 = new Readable();
      inputStream2._read = (size: number): void => {
        setImmediate(() => {
          inputStream2.emit('data', 'three');
          inputStream2.emit('data', 'four');
          inputStream2.emit('end');
        });
      };
      inputStream2.on('end', () => console.log('Input stream 2 ended'));
      mergedStream.add(inputStream2);
      mergedStream.on('finish', () => {
        console.log('merged stream ended');
        done();
      });
    });
```

What basically happens in the above example and also in the `PartialResultStream` in case of a retry is the following:
1. An initial request stream is opened (`inputStream1`) and added to the merged result stream.
2. `inputStream1` emits a retryable error. This error bubbles up to the merged stream, which intercepts it and executes a retry. Intercepting this error means that the mergedStream does not know that `inputStream1` ended with an error, and it erroneously thinks it should still receive data from it.
3. `inputStream2` is started, returns its data and then finishes. The 'end' event bubbles up to the merged stream, but that does not end the merged stream as it is still waiting for input from `inputStream1`.

The above also happened when testing `PartialResultStream` with the mock server. If you remove this `endListener` that was added, you will see that retrying errors does not actually work as intended: All data is returned to the user stream, but the stream is never ended, causing the test cases to hang.

The above example can be made to work by setting `simulateRetryAfterError = false` or by replacing the error from `inputStream1` with an end event for the input stream.

Reply:
Thanks for the detailed explanation :clap:. This means when an error happens in `inputStream1`, it will emit an `error` event, but there is no `end` event after that. The merged stream is not ended, so it will not emit `finish` event.

Reply:
FYI, when I changed your test case from `finish` to `end`, like this: 

```typescript
      mergedStream.on('end', () => {
        console.log('merged stream ended');
        done();
      });
```

I commented the error and uncommented the `end`: 

```typescript
          ...
          inputStream1.emit('end'); // Uncomment this line and comment out the next line to make the test case work.
          // inputStream1.emit('error', new Error('test error'));
          ...
```

This does not work any more. I was wondering why this happens, :sweat_smile:. Then, I found this link: https://nodejs.org/api/stream.html#stream_events_finish_and_end.

Reply:
> Thanks for the detailed explanation . This means when an error happens in `inputStream1`, it will emit an `error` event, but there is no `end` event after that. The merged stream is not ended, so it will not emit `finish` event.

Correct. And the error that is emitted by `inputStream1` is intercepted, so that also keeps the merged stream think that `inputStream1` is still alive.